File: Tests/parallel_loop_reduction_multiply_loop.c
#include "acc_testsuite.h"
#ifndef T1
//T1:parallel,loop,reduction,combined-constructs,V:1.0-2.7
int test1(){
    int err = 0;
    int m_n = 128;
    srand(SEED);
    real_t * a = (real_t *)malloc(10 * m_n * sizeof(real_t));
    real_t * b = (real_t *)malloc(10 * m_n * sizeof(real_t));
    real_t * c = (real_t *)malloc(10 * m_n * sizeof(real_t));
    real_t * totals = (real_t *)malloc(10 * sizeof(real_t));
    real_t temp;

    for (int x = 0; x < 10 * m_n; ++x){
        a[x] = .05 + rand() / (real_t)(RAND_MAX);
        b[x] = .05 + rand() / (real_t)(RAND_MAX);
        c[x] = 0.0;
    }


    #pragma acc data copyin(a[0:10*m_n], b[0:10*m_n]) copyout(c[0:10*m_n], totals[0:10])
    {
        #pragma acc parallel loop gang private(temp)
        for (int x = 0; x < 10; ++x){
            temp = 1;
            #pragma acc loop worker reduction(*:temp)
            for (int y = 0; y < m_n; ++y){
                temp *= a[x * m_n + y] + b[x * m_n + y];
            }
            totals[x] = temp;
            #pragma acc loop worker
            for (int y = 0; y < m_n; ++y){
                c[x * m_n + y] = (a[x * m_n + y] + b[x * m_n + y]) / totals[x];
            }
        }
    }

    for (int x = 0; x < 10; ++x){
        temp = 1;
        for (int y = 0; y < m_n; ++y){
            temp *= a[x * m_n + y] + b[x * m_n + y];
        }
        if (fabs(temp - totals[x]) > (temp / 2 + totals[x] / 2) * PRECISION){
            err += 1;
        }
        for (int y = 0; y < m_n; ++y){
            if (fabs(c[x * m_n + y] - ((a[x * m_n + y] + b[x * m_n + y]) / totals[x])) > PRECISION){
                err += 1;
            }
        }
    }

    return err;
}
#endif

#ifndef T2
//T2:parallel,private,reduction,combined-constructs,loop,V:2.7-2.7
int test2(){
    int err = 0;
    int m_n = 128;
    srand(SEED);
    real_t * a = (real_t *)malloc(25 * m_n * sizeof(real_t));
    real_t * b = (real_t *)malloc(25 * m_n * sizeof(real_t));
    real_t * c = (real_t *)malloc(25 * m_n * sizeof(real_t));

    real_t * totals = (real_t *)malloc(25 * sizeof(real_t));
    real_t reduced[5];

    for (int x = 0; x < n; ++x){
        a[x] = .05 + rand() / (real_t)(RAND_MAX / 10);
        b[x] = .05 + rand() / (real_t)(RAND_MAX / 10);
        c[x] = 0;
    }

    #pragma acc data copyin(a[0:25*m_n], b[0:25*m_n]) copyout(c[0:25*m_n], totals[0:25])
    {
        #pragma acc parallel loop gang private(reduced)
        for (int x = 0; x < 5; ++x) {
            for (int y = 0; y < 5; ++y) {
                reduced[y] = 1.0;
            }
            #pragma acc loop worker reduction(*:reduced)
            for (int y = 0; y < 5 * m_n; ++y) {
                reduced[y%5] *= a[x * 5 * m_n + y] + b[x * 5 * m_n + y];
            }
            for (int y = 0; y < 5; ++y) {
                totals[x * 5 + y] = reduced[y];
            }
            for (int y = 0; y < 5 * m_n; ++y) {
                c[x * 5 * m_n + y] = (a[x * 5 * m_n + y] + b[x * 5 * m_n + y]) / totals[x * 5 + (y%5)];
            }
        }
    }

    for (int x = 0; x < 5; ++x) {
        for (int y = 0; y < 5; ++y) {
            reduced[y] = 1;
        }
        for (int y = 0; y < 5 * m_n; ++y) {
            reduced[y%5] *= a[x * 5 * m_n + y] + b[x * 5 * m_n + y];
        }
        for (int y = 0; y < 5; ++y) {
            if (fabs(totals[x * 5 + y] - reduced[y]) > PRECISION) {
                err += 1;
            }
        }
        for (int y = 0; y < 5 * m_n; ++y) {
            if (fabs(c[x * 5 * m_n + y] - ((a[x * 5 * m_n + y] + b[x * 5 * m_n + y]) / totals[x * 5 + (y % 5)])) > PRECISION) {
                err += 1;
            }
        }
    }

    return err;
}
#endif

int main(){
    int failcode = 0;
    int testrun;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
#ifndef T2
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test2();
    }
    if (failed != 0){
        failcode = failcode + (1 << 1);
    }
#endif
    return failcode;
}

File: Tests/parallel_loop_reduction_multiply_loop.f90
      INTEGER FUNCTION test()
        IMPLICIT NONE
        INCLUDE "acc_testsuite.fh"
        INTEGER :: x, y !Iterators
        REAL(8),DIMENSION(10*LOOPCOUNT):: a, b, c !Data
        REAL(8),DIMENSION(10):: totals
        REAL(8) :: temp
        INTEGER :: errors = 0

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(a)
        CALL RANDOM_NUMBER(b)
        c = 0
        a = (999.4 + a) / 2000
        b = (999.4 + b) / 2000

        !$acc data copyin(a(1:10*LOOPCOUNT), b(1:10*LOOPCOUNT)) !$acc &
            !$acc copy(c(1:10*LOOPCOUNT), totals(1:10))
          !$acc parallel loop gang private(temp)
          DO x = 0, 9
            temp = 1
            !$acc loop worker reduction(*:temp)
            DO y = 1, LOOPCOUNT
                temp = temp * (a(x * LOOPCOUNT + y) + b(x * LOOPCOUNT &
                    + y))
            END DO
            totals(x + 1) = temp
            !$acc loop worker
            DO y = 1, LOOPCOUNT
                c(x * LOOPCOUNT + y) = (a(x * LOOPCOUNT + y) + b(x * &
                    LOOPCOUNT + y)) / totals(x + 1)
            END DO
          END DO
        !$acc end data

        DO x = 0, 9
          temp = 1
          DO y = 1, LOOPCOUNT
            temp = temp * (a(x * LOOPCOUNT + y) + b(x * LOOPCOUNT + y))
          END DO
            IF (abs(temp - totals(x + 1)) .gt. ((temp / 2) + &
                (totals(x + 1) / 2)) * PRECISION) THEN
            errors = errors + 1
            WRITE(*, *) temp
            WRITE(*, *) totals(x + 1)
            WRITE(*, *) " "
          END IF
          DO y = 1, LOOPCOUNT
              IF (abs(c(x * LOOPCOUNT + y) - ((a(x * LOOPCOUNT + y) + &
                  b(x * LOOPCOUNT + y)) / totals(x + 1))) .gt. c(x * &
                  LOOPCOUNT + y) * PRECISION) THEN
              errors = errors + 1
              WRITE(*, *) c(x * LOOPCOUNT + y)
              WRITE(*, *) a(x * LOOPCOUNT + y) + b(x * LOOPCOUNT + y)
              WRITE(*, *) " "
            END IF
          END DO
        END DO
        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.fh"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "OpenACC_testsuite.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 2.5 ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Test of parallel_loop_reduction_multiply_loop"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*) "Test of parallel_loop_reduction_multiply_loop"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM


